### PR TITLE
#167727730 Logout Functionality

### PR DESCRIPTION
**Pull request by dinorhythms**
#### What does this PR do?
This PR allows user to logout
#### Description of Task to be completed?
- install ioredis dependencies
- create user logout route
- create logout in user controller
- create blacklist Middleware
#### How should this be manually tested?
- Clone this branch
- Change directory into it
- Install Redis Server
- Run npm install 
- run npm run start:dev
- POST: logout using api/v1/user/logout
#### What are the relevant pivotal tracker stories?
[#167727730](https://www.pivotaltracker.com/story/show/167727730)

### CODE REVIEW

File: src/controllers/userController.js [hunk 1]
```diff
@@ -43,6 +44,22 @@ const signUp = async (req, res) => {
   }
 };
 
+const logout = async (req, res) => {
```
Comment (ladaposamuel) [inline src/controllers/userController.js]: JsDoc comment here

File: src/middlewares/blacklistMiddleware.js [hunk 2]
```diff
@@ -0,0 +1,33 @@
+import Redis from 'ioredis';
+import response from '../utils/response';
+
+const redis = new Redis(process.env.REDIS_URL)
+
+const checkBlacklist = async(req, res, next) => {
```
Comment (ladaposamuel) [inline src/middlewares/blacklistMiddleware.js]: JsDoc comment here

**Review by ladaposamuel — CHANGES_REQUESTED** [review]

**Comment by tejiri4** [comment]
Fix Hound issues and write test

File: src/middlewares/blacklistMiddleware.js [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+    
+    const token = req.headers.authorization && req.headers.authorization.split(' ')[1]
+
+    if(!token){
+        return response(res, 401, 'error', {
+            message: "Token missing, you need a token to have access"
```
Comment (tejiri4) [inline src/middlewares/blacklistMiddleware.js]: let reference our message from the message object

File: src/controllers/userController.js [hunk 4]
```diff
@@ -43,6 +44,22 @@ const signUp = async (req, res) => {
...
+
+  const redis = new Redis(process.env.REDIS_URL)
+  const token = req.headers.authorization.split(' ')[1]
+
+  try {
+    await redis.set(token, token, 'EX', 604800000)
```
Comment (tejiri4) [inline src/controllers/userController.js]: is there a reason token is used twice here?

File: src/middlewares/blacklistMiddleware.js [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+        });
+    }
+
+    try {
+        const checkedToken = await redis.get(token)
+        if(checkedToken){
+            return response(res, 500, 'error', {
+                message: 'You are already logged out!'
```
Comment (tejiri4) [inline src/middlewares/blacklistMiddleware.js]: move message to message object

**Review by tejiri4 — CHANGES_REQUESTED** [review]
Nice work kindly look into the comment drop by me

**Review by Mcdavid95 — COMMENTED** [review]
Apart from the comments and changes, it looks good. Nice work bro.

File: src/middlewares/blacklistMiddleware.js (hunk 5 above)
Comment (dinorhythms) [reply src/middlewares/blacklistMiddleware.js]: I have moved the messages to message object

File: src/controllers/userController.js (hunk 4 above)
Comment (dinorhythms) [reply src/controllers/userController.js]: I dont wont to leave the value parameter enter, the key is whats needed. Should i change the value to an empty string?

File: src/middlewares/blacklistMiddleware.js (hunk 3 above)
Comment (dinorhythms) [reply src/middlewares/blacklistMiddleware.js]: Changed to message object

File: src/middlewares/blacklistMiddleware.js (hunk 2 above)
Comment (dinorhythms) [reply src/middlewares/blacklistMiddleware.js]: JsDoc as been added

File: src/controllers/userController.js (hunk 1 above)
Comment (dinorhythms) [reply src/controllers/userController.js]: JsDoc as been added

**Comment by tejiri4** [comment]
Rebase your branch with develop

**Comment by dinorhythms** [comment]
> Rebase your branch with develop

Rebased to develop

File: .eslintrc.json [hunk 6]
```diff
@@ -7,6 +7,7 @@
     "mocha": true
   },
   "rules": {
+    "linebreak-style": 0,
```
Comment (ladaposamuel) [inline .eslintrc.json:11]: Why did you add this new rule?

Comment (dinorhythms) [reply .eslintrc.json:11]: To stop eslint linebreak errors on windows. It enforces the usage of Windows line endings: \r\n for CRLF.

Comment (ladaposamuel) [reply .eslintrc.json:11]: awesome

**Review by tejiri4 — APPROVED** [review]
LGTM
